### PR TITLE
Refactored router to implement RouterInterface

### PR DESCRIPTION
**Pull request by markshust**

### CODE REVIEW

File: Router/SpecialRouter.php [hunk 1]
```diff
@@ -29,26 +30,61 @@ class SpecialRouter extends Base
     ];
 
     /**
-     * Translate special characters in the path to names.
+     * Match a route to this router.
+     *
+     * If there is a match, replace the path with the new path.
+     * Ex. /my-module/foo-bar/baz_qux -> /my-module/foodashbar/bazunderscorequx
```
Comment (TuVanDev) [inline Router/SpecialRouter.php]: @markshust should we use `frontName` instead of `my-module`?

Comment (markshust) [reply Router/SpecialRouter.php]: Good idea -- I'll use `front-name` though to demonstrate that the dash in the front name is not replaced.

File: Router/SpecialRouter.php [hunk 2]
```diff
@@ -29,26 +30,61 @@ class SpecialRouter extends Base
...
+        $moduleName = array_shift($pathParts);
+        $pathInfo = implode('/', $pathParts);
+
+        if ($this->isMatch($pathInfo)) {
+            $newPathInfo = $this->replacePath($pathInfo);
+            $newPathInfo = "/$moduleName/$newPathInfo";
+            $request->setPathInfo($newPathInfo);
+        }
+
+        // Now that path has been updated, return null to allow the next router to match.
```
Comment (TuVanDev) [inline Router/SpecialRouter.php]: > Now that path has been updated

No, it only updates if it matches: `if ($this->isMatch($pathInfo)) {}`

Comment (TuVanDev) [reply Router/SpecialRouter.php]: @markshust

Comment (markshust) [reply Router/SpecialRouter.php]: @TuVanDev updated 👍

File: Router/SpecialRouter.php [hunk 3]
```diff
@@ -29,26 +30,61 @@ class SpecialRouter extends Base
...
+        RequestInterface $request,
+    ): ActionInterface|null {
+        $identifier = trim($request->getPathInfo(), '/');
+        $pathParts = explode('/', $identifier);
+        $moduleName = array_shift($pathParts);
+        $pathInfo = implode('/', $pathParts);
+
+        if ($this->isMatch($pathInfo)) {
+            $newPathInfo = $this->replacePath($pathInfo);
+            $newPathInfo = "/$moduleName/$newPathInfo";
```
Comment (TuVanDev) [inline Router/SpecialRouter.php]: What do you think about the following change?
current:
```
$newPathInfo = $this->replacePath($pathInfo);
$newPathInfo = "/$moduleName/$newPathInfo";
```
new:
```
$newPathInfo = "/$moduleName/" . $this->replacePath($pathInfo);
```

Comment (markshust) [reply Router/SpecialRouter.php]: Good call -- going to take it a step further though and use `sprintf`:
```
$newPathInfo = sprintf('/%s/%s', $moduleName, $this->replacePath($pathInfo));
```

**Pull request merged**